compiler/noirc_frontend/src/hir_def/types.rs

```diff
@@ -1954,13 +1990,33 @@ impl Type {
             Type::InfixExpr(lhs, op, rhs) => {
                 let infix_kind = lhs.infix_kind(&rhs);
                 if kind.unifies(&infix_kind) {
-                    let lhs_value = lhs.evaluate_to_field_element(&infix_kind)?;
-                    let rhs_value = rhs.evaluate_to_field_element(&infix_kind)?;
+                    let lhs_value =
+                        lhs.evaluate_to_field_element_helper(&infix_kind, run_simplifications)?;
+                    let rhs_value =
+                        rhs.evaluate_to_field_element_helper(&infix_kind, run_simplifications)?;
                     op.function(lhs_value, rhs_value, &infix_kind)
                 } else {
                     None
                 }
             }
+            Type::CheckedCast(to, from) => {
+                let to_value = to.evaluate_to_field_element(kind)?;
+
+                // if both 'to' and 'from' evaluate to a constant,
+                // return None unless they match
```

Comment:
I'm not sure how often they don't match, but they appear to always match in our tests.

I need to test the cases from https://github.com/noir-lang/noir/pull/6269 today, but those would be cases in which I'd expect them to not match.

There are likely under/overflow cases that we don't handle with `canonicalize`'s simplifications which could cause them to not match so I'd prefer to issue a proper error here instead of throwing a panic.